src/main/java/com/squareup/javawriter/ClassName.java

```diff
@@ -130,94 +86,100 @@ public static ClassName fromClass(Class<?> clazz) {
-   * {@link #fromClass(Class)} should be preferred as they can correctly create {@link ClassName}
+   * method assumes that the input is ASCII and follows typical Java style (lowercase package
+   * names, UpperCamelCase class names) and may produce incorrect results or throw
+   * {@link IllegalArgumentException} otherwise. For that reason, {@link #get(Class)} and
+   * {@link #get(Class)} should be preferred as they can correctly create {@link ClassName}
    * instances without such restrictions.
    */
-  public static ClassName bestGuessFromString(String classNameString) {
-    checkNotNull(classNameString);
-    List<String> parts = Splitter.on('.').splitToList(classNameString);
-    int firstClassPartIndex = -1;
-    for (int i = 0; i < parts.size(); i++) {
-      String part = parts.get(i);
+  public static ClassName bestGuess(String classNameString) {
+    String packageName = "";
+    List<String> simpleNames = new ArrayList<>();
+
+    for (String part : Splitter.on('.').split(classNameString)) {
       checkArgument(SourceVersion.isIdentifier(part));
       char firstChar = part.charAt(0);
       if (Ascii.isLowerCase(firstChar)) {
-        // looks like a package part
-        if (firstClassPartIndex >= 0) {
-          throw new IllegalArgumentException("couldn't make a guess for " + classNameString);
-        }
+        checkArgument(simpleNames.isEmpty(), "couldn't make a guess for %s", classNameString);
+        if (!packageName.isEmpty()) packageName += ".";
+        packageName += part;
       } else if (Ascii.isUpperCase(firstChar)) {
-        // looks like a class part
-        if (firstClassPartIndex < 0) {
-          firstClassPartIndex = i;
-        }
+        simpleNames.add(part);
       } else {
-        throw new IllegalArgumentException("couldn't make a guess for " + classNameString);
+        checkArgument(false, "couldn't make a guess for %s", classNameString);
       }
     }
-    int lastIndex = parts.size() - 1;
-    return new ClassName(
-        Joiner.on('.').join(parts.subList(0, firstClassPartIndex)),
-        firstClassPartIndex == lastIndex
-            ? ImmutableList.<String>of()
-            : ImmutableList.copyOf(parts.subList(firstClassPartIndex, lastIndex)),
-        parts.get(lastIndex));
+    checkArgument(!simpleNames.isEmpty(), "couldn't make a guess for %s", classNameString);
+    simpleNames.add(0, packageName);
+    return new ClassName(simpleNames);
   }
 
-  public static ClassName create(
-      String packageName, List<String> enclosingSimpleNames, String simpleName) {
-    return new ClassName(packageName, ImmutableList.copyOf(enclosingSimpleNames),
-        simpleName);
+  /**
+   * Returns a class name created from the given parts. For example, calling this with package name
+   * {@code "java.util"} and simple names {@code "Map"}, {@code "Entry"} yields {@link Map.Entry}.
+   */
+  public static ClassName get(String packageName, String... simpleNames) {
```

Comment:
Should there be a `checkArgument(simpleNames.length > 0)` to avoid calls like `ClassName.get("java.util.Map.Entry") /* oops! I actually wanted bestGuess! */`? or use a `(String packageName, String simpleName, String... simpleNames)` signature to enforce it at compile-time?